### ISSUE TITLE
[clang-tidy] false positive on cppcoreguidelines-missing-std-forward with constrained templates

### ISSUE THREAD

**Issue by rdong8** [llvm/llvm-project#180362]
```cpp
#include <concepts>

class A
{
public:
    template <std::derived_from<A> Self>
    auto operator|(this Self &&self, int) -> void {}
};

auto main() -> int
{
    A a;
    a | 1;
}
```

`Self` is a regular rvalue reference and not a forwarding reference in `A::operator|` because of the constraint (and as seen by the compiler error in `main`), but clang-tidy will still warn [cppcoreguidelines-missing-std-forward](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/missing-std-forward.html) on this.
